Allow to use AI services as tools for another AI service

mariofusco opened the pull request:
<!--
Thank you so much for your contribution!

Please fill in all the sections below.
Please open the PR as a draft initially. Once it is reviewed and approved, we will ask you to add documentation and examples.
Please note that PRs with breaking changes or without tests will be rejected.

Please note that PRs will be reviewed based on the priority of the issues they address.
We ask for your patience. We are doing our best to review your PR as quickly as possible.
Please refrain from pinging and asking when it will be reviewed. Thank you for understanding!
-->

## Issue
<!-- Please specify the ID of the issue this PR is addressing. For example: "Closes #1234" or "Fixes #1234" -->
At the moment it is not possible to use an AI service as tool for another AI service because the proxy method doesn't carry the annotation of the interface method.

## Change
<!-- Please describe the changes you made. -->
Improve discovery of the `@Tool` annotation also looking for it on the interfaces implemented by the proxy class.

## General checklist
<!-- Please double-check the following points and mark them like this: [X] -->
- [X] There are no breaking changes
- [X] I have added unit and/or integration tests for my change
- [X] The tests cover both positive and negative cases
- [X] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [ ] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green
<!-- Before adding documentation and example(s) (below), please wait until the PR is reviewed and approved. -->
- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)


## Checklist for adding new maven module
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have added my new module in the root `pom.xml` and `langchain4j-bom/pom.xml`


## Checklist for adding new embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [ ] I have added a `{NameOfIntegration}EmbeddingStoreRemovalIT` that extends from `EmbeddingStoreWithRemovalIT`

## Checklist for changing existing embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

mariofusco committed:
Allow to use AI services as tools for another AI service

glaforge commented:
Not a formal PR review but just a comment to say, YES!, I love the idea and wanted to have something like this.
It's a nice composition of AiServices to implement multi-agent systems!

glaforge commented:
One remark: it's also a pattern we see in some agent frameworks.
For example, in ADK (Agent Development Kit), you can create an "agent tool".
For (type-safe) composability of AiServices, that'll be handy, and make LangChain4j even better for multi-agent scenarios.

LizeRaes commented:
This looks great, thanks a lot and like @glaforge says, a frequently used pattern!
I have a question on context-tracking: when building generic or dynamic AiServices I struggle with tracking which AiService called which other AiService or Tool in which conversation. The tool call ID is triggered by the LLM and passes out of sight of the user. How do you deal with this? Is there a way to add tool calls to the AiService memory for example?

glaforge commented:
Excellent question @LizeRaes!

In ADK, there are different notions of [contexts](https://google.github.io/adk-docs/context/#the-different-types-of-context).
There's a tool context for when a tool is invoked, an invocation context when an agent is called, etc.
Furthermore, ADK has a concept of session, which also holds a state (a map, basically), and (sub)agents can share this state.
Additionally, there are various before/after callbacks: when an LLM is called, when a tool is called, etc.
So it's easier to track what's going on.

Now, to take an example, let's say my LLM calls a tool, in the signature of the tool method, I can add an extra (optional) parameter which is a `ToolContext`.

So maybe the definitions of our `AiService` and `Tool` could/should also add extra parameters to convey the ongoing context?

dliubarskyi commented:
Related: https://github.com/langchain4j/langchain4j/issues/1122

dliubarskyi commented on langchain4j/src/test/java/dev/langchain4j/service/common/AbstractAiServiceWithToolsIT.java:
```diff
@@ -786,4 +787,72 @@ protected void should_execute_tool_with_uuid_parameter(ChatModel model) {
+
+    @ParameterizedTest
+    @MethodSource("models")
+    void tools_as_agents_tests(ChatModel model) {
```
Please move it to the `AiServicesWithToolsIT`, as it tests not integrations but a general capability of LC4j

dliubarskyi commented on langchain4j/src/test/java/dev/langchain4j/service/common/AbstractAiServiceWithToolsIT.java:
```diff
@@ -786,4 +787,72 @@ protected void should_execute_tool_with_uuid_parameter(ChatModel model) {
+                .tools(medicalExpert, legalExpert, technicalExpert)
+                .build();
+
+        System.out.println(routerAgent.askToExpert("I broke my leg what should I do"));
```
nit: please avoid `System.out.println`

dliubarskyi commented on langchain4j/src/test/java/dev/langchain4j/service/common/AbstractAiServiceWithToolsIT.java:
```diff
@@ -786,4 +787,72 @@ protected void should_execute_tool_with_uuid_parameter(ChatModel model) {
+
+        System.out.println(routerAgent.askToExpert("I broke my leg what should I do"));
+
+        verify(medicalExpert).medicalRequest("I broke my leg what should I do");
```
nit: would also be nice to verify:
- that other experts were not called
- configure `ChatMemory` for `routerAgent` and verify the state of it after the call

dliubarskyi reviewed (commented):
@mariofusco good stuff, thank you! It would be great to have this documented 🙏

mariofusco commented:
> So maybe the definitions of our `AiService` and `Tool` could/should also add extra parameters to convey the ongoing context?

I think that the idea of a shared context between an `AiService` and the `Tool` it uses is very useful, especially if also the tools are `AiService` themselves. Specifically in this last case, in some circumstances, they could also have a shared chat memory.

For sure I will experiment with these ideas, but regarding this pull request I believe that it could be merged as it is and then the other proposed improvements could be built on top of it.

dliubarskyi commented:
Hi @mariofusco 

> For sure I will experiment with these ideas

If you do, please check discussions in https://github.com/langchain4j/langchain4j/issues/1122, there might be some insights on this topic

> but regarding this pull request I believe that it could be merged as it is and then the other proposed improvements could be built on top of it.

We could merge it as-is, but "undocumented feature is a feature which noone will use"... Another concern is that CI will become slower and more expensive because the new test will run for 13 times instead of 1

mariofusco commented:
> We could merge it as-is, but "undocumented feature is a feature which noone will use"... Another concern is that CI will become slower and more expensive because the new test will run for 13 times instead of 1

Sorry for misunderstanding, I implicitly meant "merge it as it is AFTER that I've added docs and addressed all your other remarks" :) Will do this asap.

mariofusco commented:
@dliubarskyi All done, please give it a second look.

dliubarskyi commented on docs/docs/tutorials/tools.md:
````diff
@@ -300,6 +300,78 @@ If the method has a `String` return type, the returned value is sent to the LLM
+    }
+```
+
+the `RouterAgent` can be configured to use as tools the 3 other AI services, experts in specific fields, routing the user request to one of them.
````
nit:
```suggestion
The `RouterAgent` can be configured to use as tools the 3 other AI services, experts in specific fields, routing the user request to one of them.
```

dliubarskyi commented on docs/docs/tutorials/tools.md:
```diff
@@ -300,6 +300,78 @@ If the method has a `String` return type, the returned value is sent to the LLM
+        @dev.langchain4j.service.UserMessage("""
+            Analyze the following user request and categorize it as 'legal', 'medical' or 'technical',
+            then forward the request as it is to the corresponding expert provided as a tool.
+            Finally return the answer that you received from the expert without any modification.
```
I am not sure this (routing) is a good example to include in the documentation, as users might take it as a "go-to recipe", while IMHO it has some big drawbacks.

This implementation requires the LLM to copy-paste the user request (can be very long) without modifications as a tool call, then copy-paste the expert's response (can be very long) without modifications  to the user, so it:
1. consumes way more tokens than it could (latency + money)
2. error-prone because of copy-pasting

Perhaps you have other ideas what can be a good example?

mariofusco replied:
I see your point and this is exactly what I discussed here https://quarkus.io/blog/agentic-ai-with-quarkus-p3/#comparing-the-workflow-and-agentic-approaches

However this is quite unavoidable at the moment and I cannot think of any other example that wouldn't suffer of this problem. The tool result is always reprocessed by the LLM who called it, regardless if also the tool is an LLM. The only structural solution to this that I see will be merging that other pull request allowing to directly return the tool's response.

dliubarskyi replied:
I see... returning tool directly can help, but only partially, as input should be copy-pasted. Unless the agent-tool accepts processed input (e.g., only a relevant part).

Another concern is that follow-up questions to the agent-tool will lack the memory, unless we explicitly instruct the main agent to include memory, but it is brittle.

Perhaps worth adding a note/arning into the documentation about the limitations of this approach?

mariofusco replied:
Ok, good point. I will clarify all those limitations in docs.

mariofusco replied:
I added to docs a note reporting the issues that we discussed, please review again.

dliubarskyi commented on langchain4j/src/test/java/dev/langchain4j/service/AiServicesWithToolsIT.java:
```diff
@@ -862,4 +862,82 @@ LocalDate currentDate() {
+
+        routerAgent.askToExpert("I broke my leg what should I do");
+        verify(medicalExpert).medicalRequest("I broke my leg what should I do");
+        verifyNoMoreInteractions(legalExpert, technicalExpert);
```
```suggestion
        verifyNoInteractions(legalExpert, technicalExpert);
```

dliubarskyi reviewed (commented):
@mariofusco thanks a lot!

kpavlov commented on docs/docs/tutorials/tools.md:
```diff
@@ -300,6 +300,85 @@ If the method has a `String` return type, the returned value is sent to the LLM
+    interface MedicalExpert {
+
+        @dev.langchain4j.service.UserMessage("""
+            You are a medical expert.
```
Nit: I would extract system prompt part to separate annotation. In a separate PR

kpavlov commented on langchain4j-core/src/main/java/dev/langchain4j/internal/Utils.java:
```diff
@@ -396,4 +400,34 @@ public static Map<String, String> toStringValueMap(Map<String, Object> map) {
+     * @return An {@link Optional} containing the method having the given annotation,
+     *         or an empty {@link Optional} if there isn't any.
+     */
+    public static Optional<Method> getAnnotatedMethod(Method method, Class<? extends Annotation> annotation) {
```
nit: would be great to update UtilsTest

mariofusco replied:
Good suggestion, done.

kpavlov reviewed (dismissed):
Thank you, @mariofusco ! Very useful feature. Make sense to merge as it is

dliubarskyi reviewed (approved):
@mariofusco thanks a lot!
